### PR TITLE
Mvp 1306 update neo node labelling generation script to ensure no duplicates

### LINKED ISSUE #41: Update neo_node_labelling generation script to ensure no duplicates
The script as currently structured would duplicate entries if run twice on the same file.  

TODO: UPdate the script to avoid duplicate entries:

Option1: Use set + collections.named_tuple(); Option2:  Transpose the data structure to a dictionary with label as key. Update dict then transpose back and update YAML.  The latter imposes a restriction that labels must be unique, but still allows for multiple DL queries to generate the same label as 'classes' is a list.  

Also - please refactor so that updating neo_node_labelling is a function.

### PR DESCRIPTION
**Pull request by ubyndr**
Fixes #41 , Used namedtuple to prevent duplicates in neo_node_labelling section.

### CODE REVIEW

File: src/write_organ_tags.py [hunk 1]
```diff
@@ -6,9 +6,77 @@
 #  Output will be used to filter and/or boost search results in Cell Annotation Platform (CAP)
 
 import argparse
-import sys
 import ruamel.yaml
 from SPARQLWrapper import SPARQLWrapper, JSON
+from collections import namedtuple
+
+
```
Comment (evanbiederstedt) [inline src/write_organ_tags.py:13]: I think most of these functions are clear to me. However, could you add 2-3 sentences of explanatory comments for each?

You could put these in `# comments` or in 
```
"""
comments
"""
```

File: src/write_organ_tags.py [hunk 2]
```diff
@@ -6,9 +6,77 @@
...
+        output.append(line)
+    return output
+
+
+def generate_organ_cells(output):
+    organs = []
+    for n in output:
+        curie = n['x']['value'].replace("_", ":")
+        curie = curie.partition('http://purl.obolibrary.org/obo/')[-1]
+        dl_query = "CL:0000000 and BFO:0000050 some " + curie
```
Comment (evanbiederstedt) [inline src/write_organ_tags.py:28]: As an example of a place whereby I could use a comment, here is where an explanation is necessary:

`dl_query = "CL:0000000 and BFO:0000050 some " + curie`

Why both CL and BFO?

**Review by evanbiederstedt — CHANGES_REQUESTED** [review]
Request above to add more explanatory details to the functions. That would make the following the logic of what's going on here easier for me. 

RE: namedtuple

We're doing this as opposed to a dictionary to ensure that the keys are strings?

File: src/write_organ_tags.py (hunk 2 above)
Comment (ubyndr) [reply src/write_organ_tags.py:28]: I think this is trying to say things that are 'Cell', CL:0000000, and 'part of', BFO:0000050, some specific organs. Make sense?

Comment (dosumis) [reply src/write_organ_tags.py:28]: It is good practice to write out queries using labels in comments.

File: src/write_organ_tags.py (hunk 1 above)
Comment (dosumis) [reply src/write_organ_tags.py:13]: Well formatted docstrings for each function would be best.  I'd prefer these not to be treated as comments, but to describe args, output and what the function does.  Code comments are great, but better not to put them in docstrings.

**Comment by dosumis** [comment]
> RE: namedtuple
> We're doing this as opposed to a dictionary to ensure that the keys are strings?

It's hashable (as long as its component parts are) so can be used with set() to remove duplicates.

**Review by dosumis — CHANGES_REQUESTED** [review]
* Please refactor to move generic functions to a separate file (bonus points for using python packaging).  This will allow us to re-use these functions for other automated extensions
* Please add docstrings to functions (especially important for the generic ones you move to a separate file).

**Comment by ubyndr** [comment]
I've tried follow this guide to create the python package, https://packaging.python.org/en/latest/tutorials/packaging-projects/. All the names are horrible I guess, open for suggestions. Also I have uploaded the package src code to here for testing purposes, https://github.com/ubyndr/config_utils, also open for suggestions where to keep it. @dosumis can you please review it again?

**Comment by evanbiederstedt** [comment]
I think it's standard to put the argparse section at the top of the script, or within a `main()` function to be run. 

That is, this part:

```
parser = argparse.ArgumentParser(description = 'set destination YAML file for query output')

parser.add_argument('-f', '--file', default = '../config/prod/neo4j2owl-config.yaml', help = '''
    Use this option to indicate destination file for organ cell DL queries and semantic labels. By default, output
    is sent to a file named neo4j2owl-config.yaml.
    ''')

args = parser.parse_args()

file_name = args.file
```

That said, it's not a big deal either way.

**Review by evanbiederstedt — APPROVED** [review]

**Comment by ubyndr** [comment]
I have changed the package name and upload it to https://pypi.org/project/cap-pipeline-config-utils/0.0.1/

**Comment by dosumis** [comment]
Was just expecting local package rather than something on PyPi. Sorry if not clear.  Code here looks good though: https://github.com/ubyndr/config_utils/blob/main/src/config_autogenerate_utils/utils.py 

If keeping as as package on PyPi, we need to think about a better home for the code.  One option is to package with the JSON schema and schema checker.  I like the idea of utils including schema update and schema validation.

More general discussion needed on factoring out this pipeline as a product distinct from VFB.

**Pull request merged**